### ISSUE TITLE
manual installation process on ubuntu with TPM error

### ISSUE THREAD

**Issue by franck-serot** [keylime/rust-keylime#597]
Hi,
After reading the keylime documentation, I try to install the **keylime** project and the **rust-keylime** project on an ubuntu virtual machine 22.04 LTS with an emulated TPM v2.0.
I identified and solved a dependencies problem for the **rust-keylime** project.
I have a TPM error after creating of an primary object with the **rust-keylime** project.
But, I can directly create an primary object.
Could you help me to confirm the installation process below ?
Could you try to reproduce the TPM error and help me solve it ?
Best regards,
Franck

- create an ubuntu virtual machine 22.04 LTS with
   - a proxy configuration
   - a Python configuration about repository
   - Rust environnement
- add an emulated TPM v2.0 hardware on the previous virtual machine
- install the **keylime** project v7.0.0
```
$ mkdir ~/project
$ cd ~/project
$ sudo apt install git
$ git clone https://github.com/keylime/keylime.git
$ cd ~/project/keylime
$ sudo -E -s
# ./install.sh
```
- start the components **keylime_verifier** and **keylime_registrar** in the 2 shells
```
# keylime_verifier
...
```
```
# keylime_registrar
...
```
- check the status of the 2 previous components with the **keylime_tenant** component in another shell
```
# keylime_tenant -c cvstatus
INFO:keylime.config:Reading configuration from ['/etc/keylime/logging.conf']
2023-05-17 14:35:06.164 - keylime.tpm - INFO - TPM2-TOOLS Version: 5.2
2023-05-17 14:35:06.166 - keylime.config - INFO - Reading configuration from ['/etc/keylime/tenant.conf']
2023-05-17 14:35:06.167 - keylime.tenant - INFO - Setting up client TLS...
2023-05-17 14:35:06.167 - keylime.tenant - INFO - Using default client_cert option for tenant
2023-05-17 14:35:06.167 - keylime.tenant - INFO - Using default client_key option for tenant
2023-05-17 14:35:06.167 - keylime.tenant - INFO - No value provided in client_key_password option for tenant, assuming the key is unencrypted
2023-05-17 14:35:06.168 - keylime.tenant - INFO - TLS is enabled.
2023-05-17 14:35:06.168 - keylime.tenant - WARNING - Using default UUID d432fbb3-d2f1-4a97-9ef7-75bd81c00000
2023-05-17 14:35:06.168 - keylime.tenant - INFO - Verifier at 127.0.0.1 with Port 8881 does not have agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000.
 
# keylime_tenant -c regstatus
INFO:keylime.config:Reading configuration from ['/etc/keylime/logging.conf']
2023-05-17 14:35:18.939 - keylime.tpm - INFO - TPM2-TOOLS Version: 5.2
2023-05-17 14:35:18.941 - keylime.config - INFO - Reading configuration from ['/etc/keylime/tenant.conf']
2023-05-17 14:35:18.941 - keylime.tenant - INFO - Setting up client TLS...
2023-05-17 14:35:18.941 - keylime.tenant - INFO - Using default client_cert option for tenant
2023-05-17 14:35:18.941 - keylime.tenant - INFO - Using default client_key option for tenant
2023-05-17 14:35:18.941 - keylime.tenant - INFO - No value provided in client_key_password option for tenant, assuming the key is unencrypted
2023-05-17 14:35:18.943 - keylime.tenant - INFO - TLS is enabled.
2023-05-17 14:35:18.943 - keylime.tenant - WARNING - Using default UUID d432fbb3-d2f1-4a97-9ef7-75bd81c00000
2023-05-17 14:35:18.993 - keylime.registrar_client - CRITICAL - Error: could not get agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 data from Registrar Server: 404
2023-05-17 14:35:18.993 - keylime.registrar_client - CRITICAL - Response code 404: agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 not found
2023-05-17 14:35:18.994 - keylime.tenant - INFO - Agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 does not exist on the registrar. Please register the agent with the registrar.
2023-05-17 14:35:18.994 - keylime.tenant - INFO - {"code": 404, "status": "Agent d432fbb3-d2f1-4a97-9ef7-75bd81c00000 does not exist on registrar 127.0.0.1 port 8891.", "results": {}}
```
- install the **rust-keylime** projet v0.2.1
```
$ cd ~/project
$ git clone https://github.com/keylime/rust-keylime.git
$ cd ~/project/rust-keylime
$ sudo apt install libssl-dev gcc libtss-dev libzmq3-dev libarchive-dev
!!! additional dependencies !!!
$ sudo apt install pkg-config libtss2-dev libclang-15-dev
$ sudo adduser --system --shell /bin/false --home /var/lib/keylime --no-create-home keylime
$ sudo mkdir -p /var/lib/keylime
$ sudo chown keylime -R /var/lib/keylime
$ sudo -E -s
# RUST_LOG=keylime_agent=trace cargo run --bin keylime_agent
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/keylime_agent`
 DEBUG keylime_agent::secure_mount > Secure store location /var/lib/keylime/secure already mounted on tmpfs
 INFO  keylime_agent::permissions  > Changed file /var/lib/keylime/secure owner to keylime:tss.
 INFO  keylime_agent::permissions  > Dropped privileges to run as keylime:tss
 INFO  keylime_agent               > Running the service as keylime:tss...
 INFO  keylime_agent               > Starting server with API version v2.0...
 WARN  keylime_agent               > INSECURE: Keylime is using a software TPM emulator rather than a real hardware TPM.
 WARN  keylime_agent               > INSECURE: The security of Keylime is NOT linked to a hardware root of trust.
 WARN  keylime_agent               > INSECURE: Only use Keylime in this mode for testing or debugging purposes.
WARNING:esys:src/tss2-esys/api/Esys_CreatePrimary.c:400:Esys_CreatePrimary_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/api/Esys_CreatePrimary.c:135:Esys_CreatePrimary() Esys Finish ErrorCode (0x000009a2)
Error: Tpm(Tss2 { err: Tss2Error(FormatOne(FormatOneResponseCode { .0: 2466, error_number: 34, parameter: false, format_selector: true, number: 9 })), kind: Some(BadAuth), message: "authorization failure without DA implications (associated with session number 1)" })
```
- try to Create an ECC primary object (cf https://github.com/tpm2-software/tpm2-tools/blob/master/man/tpm2_createprimary.1.md)
```
$ tpm2_createprimary -C o -g sha256 -G ecc -c context.out
name-alg:
  value: sha256
  raw: 0xb
attributes:
  value: fixedtpm|fixedparent|sensitivedataorigin|userwithauth|restricted|decrypt
  raw: 0x30072
type:
  value: ecc
  raw: 0x23
curve-id:
  value: NIST p256
  raw: 0x3
kdfa-alg:
  value: null
  raw: 0x10
kdfa-halg:
  value: (null)
  raw: 0x0
scheme:
  value: null
  raw: 0x10
scheme-halg:
  value: (null)
  raw: 0x0
sym-alg:
  value: aes
  raw: 0x6
sym-mode:
  value: cfb
  raw: 0x43
sym-keybits: 128
x: a510578a106eb4fe82ab35a5dfaab23f0032d7a0cc5e7a9a6a692705d413fff4
y: 6546b393a37a0b81044f87d2b52c519148cecf946f12e54f52596eb6adbe8313
```

**Comment by ansasaki** [comment]
@franck-serot From the error message, I think probably you are setting a `tpm_ownerpassword` on `/etc/keylime/agent.conf` and the TPM owner password was not configured (by default it is empty). If you configured the TPM owner password, then you should set the correct password in the `tpm_ownerpassword` option.

Another issue is that ECC key is not supported at the moment, please use RSA.

**Comment by franck-serot** [comment]
@ansasaki Thank you for your respons

This is my contribution about the manuel instalaltion process.
This issue can be closed.

The keylime project introduces the agent configuration file in /etc/keylime/agent.conf containing the parameter 'tpm_ownerpassword = "keylime"'.
Moreover, the header of the /etc/keylime/agent.conf file underlines the obsolescence of this file.
```
# Keylime agent configuration
# The Python agent is deprecated and will be removed with the next major release (7.0.0)!
# Please migrate to the Rust based agent: https://github.com/keylime/rust-keylime/
```
 
The rust-keylime project introduces the new agent configuration file ~/project/rust-keylime/keylime-agent.conf containing the parameter 'tpm_ownerpassword = ""'.
We must therefore replace the content of the /etc/keylime/agent.conf file with that of the ~/project/rust-keylime/keylime-agent.conf file before starting the keylime agent.
If the 'tpm_ownerpassword' parameter is initialized with a non-empty value, then the authorization values ​​for the TPM objects must also be changed with the 'tpm2_changeauth' command

[set-owner-endorsement-and-lockout-authorizations-to-newpass](https://github.com/tpm2-software/tpm2-tools/blob/master/man/tpm2_changeauth.1.md#set-owner-endorsement-and-lockout-authorizations-to-newpass)
```
$ tpm2_changeauth -c owner newpass
$ tpm2_changeauth -c endorsement newpass
$ tpm2_changeauth -c lockout newpass
```
